mono/metadata/profiler.c

```diff
@@ -772,3 +785,78 @@ mono_profiler_set_events (int flags)
 {
 	/* Do nothing. */
 }
+
+static void
+allocation_cb (MonoProfiler *prof, MonoObject* object)
+{
+	prof->allocation (prof->profiler, object, object->vtable->klass);
+}
+
+void
+mono_profiler_install_allocation (MonoLegacyProfileAllocFunc callback)
+{
+	current->allocation = callback;
+
+	if (callback)
+		mono_profiler_set_gc_allocation_callback (current->handle, allocation_cb);
+}
+
+static void
+enter_cb (MonoProfiler *prof, MonoMethod *method, MonoProfilerCallContext *context)
+{
+	prof->enter (prof->profiler, method);
+}
+
+static void
+leave_cb (MonoProfiler *prof, MonoMethod *method, MonoProfilerCallContext *context)
+{
+	prof->leave (prof->profiler, method);
+}
+
+void
+mono_profiler_install_enter_leave (MonoLegacyProfileMethodFunc enter, MonoLegacyProfileMethodFunc fleave)
+{
+	current->enter = enter;
+	current->leave = fleave;
+
+	if (enter)
+		mono_profiler_set_method_enter_callback (current->handle, enter_cb);
+
+	if (fleave)
+		mono_profiler_set_method_leave_callback (current->handle, leave_cb);
```

Comment:
This isn't enough: You'd also need to install a tail call event callback as the new API separates those two cases.